spring-data-jdbc/src/main/java/org/springframework/data/jdbc/repository/aot/JdbcCodeBlocks.java

```diff
@@ -0,0 +1,789 @@
+							typeToRead);
+				}
+
+				if (StringUtils.hasText(resultSetExtractorRef) || resultSetExtractorClass != null) {
+
+					resultSetExtractor = context.localVariable("resultSetExtractor");
+
+					if (resultSetExtractorClass != null && (rowMapperClass != null || StringUtils.hasText(rowMapperRef))) {
+						builder.addStatement("$T $L = new $T($L)", ResultSetExtractor.class, resultSetExtractor,
+								resultSetExtractorClass, rowMapper);
+					} else if (resultSetExtractorClass != null) {
+						builder.addStatement("$T $L = new $T()", ResultSetExtractor.class, resultSetExtractor,
+								resultSetExtractorClass);
+					} else if (StringUtils.hasText(resultSetExtractorRef)) {
+						builder.addStatement("$T $L = getRowMapperFactory().getResultSetExtractor($S)", ResultSetExtractor.class,
+								resultSetExtractor, resultSetExtractorRef);
+					}
+				}
+
+				if (StringUtils.hasText(resultSetExtractor)) {
+
+					builder.addStatement("return ($T) getJdbcOperations().query($L, $L, $L)", queryResultType, queryVariableName,
+							parameterSourceVariableName, resultSetExtractor);
+
+					return builder.build();
+				}
+
+				boolean dynamicProjection = StringUtils.hasText(context.getDynamicProjectionParameterName());
+				Object queryResultTypeRef = dynamicProjection ? context.getDynamicProjectionParameterName() : queryResultType;
+
+				if (queryMethod.isCollectionQuery() || queryMethod.isSliceQuery() || queryMethod.isPageQuery()) {
+
+					builder.addStatement("$1T $2L = ($1T) getJdbcOperations().query($3L, $4L, new $5T<>($6L))", List.class,
+							result, queryVariableName, parameterSourceVariableName, RowMapperResultSetExtractor.class, rowMapper);
+
+					if (queryMethod.isSliceQuery() || queryMethod.isPageQuery()) {
+
+						String pageable = context.getPageableParameterName();
+
+						builder.addStatement(
+								"$1T $2L = ($1T) convertMany($3L, %s)".formatted(dynamicProjection ? "$4L" : "$4T.class"), List.class,
+								context.localVariable("converted"), result, queryResultTypeRef);
+
+						if (queryMethod.isPageQuery()) {
+
+							builder.addStatement("return $1T.getPage($2L, $3L, $4L)", PageableExecutionUtils.class,
+									context.localVariable("converted"), pageable, context.localVariable("countAll"));
+						} else {
+
+							builder.addStatement("boolean $1L = $2L.isPaged() && $3L.size() > $2L.getPageSize()",
+									context.localVariable("hasNext"), pageable, context.localVariable("converted"));
+
+							builder.addStatement("return new $1T($2L ? $3L.subList(0, $4L.getPageSize()) : $3L, $4L, $2L)",
+									SliceImpl.class, context.localVariable("hasNext"), context.localVariable("converted"), pageable);
+						}
+
+						return builder.build();
+					}
+
+					builder.addStatement("return ($T) convertMany($L, %s)".formatted(dynamicProjection ? "$L" : "$T.class"),
+							context.getReturnTypeName(), result, queryResultTypeRef);
```

Comment:
should we use string formatting to bypass more expressive use of the builder API? Puts additional cognitive load on the reader having to deal with 2 placeholder patterns and a dynamic result type ref.